### PR TITLE
Implement simple match extend

### PR DESCRIPTION
**Pull request by adraismawur**
This implements the simple match extend function, which ignores the idea of mismatching and should be much easier to explain than the legacy extend.

### CODE REVIEW

File: big_scape/comparison/extend.py [hunk 1]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+    )
+
+    # TODO: this is obviously repetitive, but beyond extracting a function I'm
+    # not sure how to make it better
+    # for now I'd prefer this to be obvious and clear over looking good
+
+    # a forward
+    score = 0
+    max_score = 0
+    for i in range(pair.comparable_region.lcs_a_stop + 1, len(a_domains)):
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:635]: lcs_a_stop governs cds bounds, should be lcs_domain_a_stop.
comparable region stops are exclusive, so remove the +1
```suggestion
    for i in range(pair.comparable_region.lcs_domain_a_stop, len(a_domains)):
```

File: big_scape/comparison/extend.py [hunk 2]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+        if domain not in common_domains:
+            score += gap
+        else:
+            score += match
+
+        if score > max_score:
+            cds_idx = a_domains[i][1]
+            max_score = score
+            pair.comparable_region.a_stop = cds_idx
+            pair.comparable_region.domain_a_stop = i
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:647]: stops should be exclusive, so +1 to both cds and domain stops

File: big_scape/comparison/extend.py [hunk 3]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+
+    # a reverse
+    score = 0
+    max_score = 0
+    for i in range(pair.comparable_region.lcs_a_start - 1, -1, -1):
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:652]:
```suggestion
    for i in range(pair.comparable_region.lcs_domain_a_start - 1, -1, -1):
```

File: big_scape/comparison/extend.py [hunk 4]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+        if score > max_score:
+            cds_idx = a_domains[i][1]
+            max_score = score
+            pair.comparable_region.a_start = cds_idx
+            pair.comparable_region.domain_a_start = i
+
+    # b forward
+    score = 0
+    max_score = 0
+    for i in range(pair.comparable_region.lcs_b_stop + 1, len(b_domains)):
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:669]:
```suggestion
    for i in range(pair.comparable_region.lcs_domain_b_stop, len(b_domains)):
```

File: big_scape/comparison/extend.py [hunk 5]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+        if domain not in common_domains:
+            score += gap
+        else:
+            score += match
+
+        if score > max_score:
+            cds_idx = b_domains[i][1]
+            max_score = score
+            pair.comparable_region.b_stop = cds_idx
+            pair.comparable_region.domain_b_stop = i
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:681]: add +1's

File: big_scape/comparison/extend.py [hunk 6]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+
+    # b reverse
+    score = 0
+    max_score = 0
+    for i in range(pair.comparable_region.lcs_b_start - 1, -1, -1):
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:686]:
```suggestion
    for i in range(pair.comparable_region.lcs_domain_b_start - 1, -1, -1):
```

File: test/comparison/test_extend.py [hunk 7]
```diff
@@ -882,3 +882,54 @@ def test_expand_greedy(self):
...
+        """
+
+        a_cds, b_cds = generate_mock_cds_lists(14, 10, [6, 9, 10], [5, 6, 7], False)
+        record_a = generate_mock_region(a_cds)
+        record_b = generate_mock_region(b_cds)
+        pair = big_scape.comparison.record_pair.RecordPair(record_a, record_b)
+
+        # emulate LCS
+        pair.comparable_region = bs_comp.ComparableRegion(
+            9, 10, 6, 7, 9, 10, 6, 7, False
```
Comment (nlouwen) [inline test/comparison/test_extend.py:910]:
```suggestion
            9, 11, 6, 8, 9, 11, 6, 8, False
```
exclusive ends

File: test/comparison/test_extend.py [hunk 8]
```diff
@@ -882,3 +882,54 @@ def test_expand_greedy(self):
...
+        )
+
+        bs_comp.extend.extend_simple_match(pair, 5, -2)
+
+        expected_comparable_region = bs_comp.ComparableRegion(
+            6, 10, 5, 7, 6, 10, 5, 7, False
```
Comment (nlouwen) [inline test/comparison/test_extend.py:916]:
```suggestion
            6, 11, 5, 8, 6, 11, 5, 8, False
```

File: test/comparison/test_extend.py [hunk 9]
```diff
@@ -882,3 +882,54 @@ def test_expand_greedy(self):
...
+            expected_comparable_region.domain_b_start,
+        )
+        self.assertEqual(
+            pair.comparable_region.domain_a_stop,
+            expected_comparable_region.domain_a_stop,
+        )
+        self.assertEqual(
+            pair.comparable_region.domain_b_stop,
+            expected_comparable_region.domain_b_stop,
+        )
```
Comment (nlouwen) [inline test/comparison/test_extend.py:935]: To include if you want, a test with multidomain cds, I used it to figure out the exclusive end confusions

```suggestion
        )
    def test_expand_simple_match_multi_domain(self):
        """Tests glocal expand with multi domain cdss"""
        # brackets indicate a cds with multiple domains
        #
        # A:        [XX]BX[A BC]DEX XXXX
        # B: XXXXXXXXXX XX A[BC]EDX[XXXX]
        #
        a_cds, b_cds = generate_mock_cds_lists(
            10, 17, [3, 3, 3, 4, 5], [12, 13, 13, 15, 14], False
        )
        a_cds[0].hsps.append(a_cds[0].hsps[0])
        a_cds[1].hsps = [a_cds[3].hsps[1]]
        b_cds[-1].hsps.extend([b_cds[-1].hsps[0]] * 3)

        record_a = generate_mock_region(a_cds)
        record_b = generate_mock_region(b_cds)
        pair = big_scape.comparison.record_pair.RecordPair(record_a, record_b)
        pair.comparable_region = bs_comp.ComparableRegion(
            3, 4, 12, 14, 4, 7, 12, 15, False
        )
        bs_comp.extend.extend_simple_match(pair, 5, -2)
        expected_comparable_region = bs_comp.ComparableRegion(
            1, 6, 12, 16, 2, 9, 12, 17, False
        )

        self.assertEqual(pair.comparable_region, expected_comparable_region)
        self.assertEqual(
            pair.comparable_region.domain_a_start,
            expected_comparable_region.domain_a_start,
        )
        self.assertEqual(
            pair.comparable_region.domain_b_start,
            expected_comparable_region.domain_b_start,
        )
        self.assertEqual(
            pair.comparable_region.domain_a_stop,
            expected_comparable_region.domain_a_stop,
        )
        self.assertEqual(
            pair.comparable_region.domain_b_stop,
            expected_comparable_region.domain_b_stop,
        )
```

**Review by nlouwen — CHANGES_REQUESTED** [review]
Overall logic is sound, some indexing discrepancies if I am not mistaken. Also needs a merge from dev to add simple_match to the new EXTEND_STRATEGY enum, so mac can multiprocess

File: big_scape/comparison/extend.py [hunk 10]
```diff
@@ -592,3 +592,107 @@ def extend_greedy(pair: RecordPair) -> None:
...
+    # I *think* the simplest way to do this is to create a list of domains with their source cds. for reasons.
+    a_domains = []
+    b_domains = []
+
+    # so we'll do a loop through cds and through domains to keep track of everything
+    for cds_idx, cds in enumerate(pair.record_a.get_cds()):
+        for domain in cds.hsps:
+            a_domains.append((domain, cds_idx))
+
+    for cds_idx, cds in enumerate(pair.record_b.get_cds()):
```
Comment (nlouwen) [inline big_scape/comparison/extend.py:619]:
```suggestion
    for cds_idx, cds in enumerate(pair.record_b.get_cds_with_domains()):
```

**Comment by adraismawur** [comment]
Excellent. As usual, working with the coordinates and the cds/domain distinction is not my strong suit.

**Review by nlouwen — APPROVED** [review]
lgtm!

**Pull request merged**